client/src/client.ts

```diff
@@ -0,0 +1,2022 @@
+	}
+
+	public start(): Disposable {
+		this._listeners = [];
+		this._providers = [];
+		// If we restart then the diagnostics collection is reused.
+		if (!this._diagnostics) {
+			this._diagnostics = this._clientOptions.diagnosticCollectionName
+				? Languages.createDiagnosticCollection(this._clientOptions.diagnosticCollectionName)
+				: Languages.createDiagnosticCollection();
+		}
+
+		this.state = ClientState.Starting;
+		this.resolveConnection().then((connection) => {
+			connection.onLogMessage((message) => {
+				switch (message.type) {
+					case MessageType.Error:
+						this.error(message.message);
+						break;
+					case MessageType.Warning:
+						this.warn(message.message);
+						break;
+					case MessageType.Info:
+						this.info(message.message);
+						break;
+					default:
+						this.outputChannel.appendLine(message.message);
+				}
+			});
+			connection.onShowMessage((message) => {
+				switch (message.type) {
+					case MessageType.Error:
+						Window.showErrorMessage(message.message);
+						break;
+					case MessageType.Warning:
+						Window.showWarningMessage(message.message);
+						break;
+					case MessageType.Info:
+						Window.showInformationMessage(message.message);
+						break;
+					default:
+						Window.showInformationMessage(message.message);
+				}
+			});
+			connection.onRequest(ShowMessageRequest.type, (params) => {
+				let messageFunc: <T extends MessageItem>(message: string, ...items: T[]) => Thenable<T>;
+				switch (params.type) {
+					case MessageType.Error:
+						messageFunc = Window.showErrorMessage;
+						break;
+					case MessageType.Warning:
+						messageFunc = Window.showWarningMessage;
+						break;
+					case MessageType.Info:
+						messageFunc = Window.showInformationMessage;
+						break;
+					default:
+						messageFunc = Window.showInformationMessage;
+				}
+				let actions = params.actions || [];
+				return messageFunc(params.message, ...actions);
+			});
+			connection.onTelemetry((data) => {
+				this._telemetryEmitter.fire(data);
+			});
+			this.initRegistrationHandlers(connection);
+			connection.listen();
+			// Error is handled in the intialize call.
+			this.initialize(connection).then(undefined, () => { });
+		}, (error) => {
+			this.state = ClientState.StartFailed;
+			this._onReadyCallbacks.reject(error);
+			this.error('Starting client failed', error);
+			Window.showErrorMessage(`Couldn't start client ${this._name}`);
+		});
+		return new Disposable(() => {
+			if (this.needsStop()) {
+				this.stop();
+			}
+		});
+	}
+
+	private resolveConnection(): Thenable<IConnection> {
+		if (!this._connectionPromise) {
+			this._connectionPromise = this.createConnection();
+		}
+		return this._connectionPromise;
+	}
+
+	private initialize(connection: IConnection): Thenable<InitializeResult> {
+		this.refreshTrace(connection, false);
+		let initOption = this._clientOptions.initializationOptions;
+		let initParams: InitializeParams = {
+			processId: process.pid,
```

Comment:
i am very much in favour of this PR.
for the tests, maybe, try to bootup the new client in the browser so see if it loads.
that way you might be able to errors like the above fairly early.